agent/peering_endpoint_test.go

```diff
@@ -12,19 +13,215 @@ import (
+	})
+
+	type testcase struct {
+		agent     *TestAgent
+		peerName  string
+		prevCount int
+	}
+
+	checkPeeringList := func(t *testing.T, a *TestAgent, expect int) {
+		req, err := http.NewRequest("GET", "/v1/peerings", nil)
+		require.NoError(t, err)
+
+		resp := httptest.NewRecorder()
+		a.srv.h.ServeHTTP(resp, req)
+		require.Equal(t, http.StatusOK, resp.Code)
+
+		var apiResp []*api.Peering
+		require.NoError(t, json.NewDecoder(resp.Body).Decode(&apiResp))
+
+		require.Len(t, apiResp, expect)
+	}
+
+	testConn := func(t *testing.T, conn *grpc.ClientConn, peers map[string]int) {
+		t.Cleanup(func() { conn.Close() })
+
+		rpcClientPeering := pbpeering.NewPeeringServiceClient(conn)
+
+		peer := &gpeer.Peer{}
+		_, err := rpcClientPeering.PeeringList(
+			context.Background(),
+			&pbpeering.PeeringListRequest{},
+			grpc.Peer(peer),
+		)
+		require.NoError(t, err)
+
+		peers[peer.Addr.String()]++
+	}
+
+	var (
+		standardPeers = make(map[string]int)
+		leaderPeers   = make(map[string]int)
+	)
+	runOnce := func(t *testing.T, tc testcase) {
+		conn, err := tc.agent.baseDeps.GRPCConnPool.ClientConn("dc1")
+		require.NoError(t, err)
+		testConn(t, conn, standardPeers)
+
+		leaderConn, err := tc.agent.baseDeps.GRPCConnPool.ClientConnLeader()
+		require.NoError(t, err)
+		testConn(t, leaderConn, leaderPeers)
+
+		checkPeeringList(t, tc.agent, tc.prevCount)
+
+		body := &pbpeering.GenerateTokenRequest{
+			PeerName: tc.peerName,
+		}
+
+		bodyBytes, err := json.Marshal(body)
+		require.NoError(t, err)
+
+		req, err := http.NewRequest("POST", "/v1/peering/token", bytes.NewReader(bodyBytes))
+		require.NoError(t, err)
+
+		resp := httptest.NewRecorder()
+		tc.agent.srv.h.ServeHTTP(resp, req)
+		require.Equal(t, http.StatusOK, resp.Code, "expected 200, got %d: %v", resp.Code, resp.Body.String())
+
+		var r pbpeering.GenerateTokenResponse
+		require.NoError(t, json.NewDecoder(resp.Body).Decode(&r))
+
+		checkPeeringList(t, tc.agent, tc.prevCount+1)
+	}
+
+	// Try the procedure on all agents to force N-1 of them to leader-forward.
+	cases := []testcase{
+		{agent: a1, peerName: "peer-1", prevCount: 0},
+		{agent: a2, peerName: "peer-2", prevCount: 1},
+		{agent: a3, peerName: "peer-3", prevCount: 2},
+	}
+
+	for i, tc := range cases {
+		tc := tc
+		testutil.RunStep(t, "server-"+strconv.Itoa(i+1), func(t *testing.T) {
+			runOnce(t, tc)
+		})
+	}
+
+	t.Run("ensure we got the right mixture of responses", func(t *testing.T) {
+		// Each read connected locally initially, so these are equally
+		// distributed with one hit each.
+		assert.Len(t, standardPeers, 3)
+		for p, n := range standardPeers {
+			assert.Equal(t, 1, n, "peer %q expected one use", p)
+		}
+
+		// Each server talks to a single leader.
+		assert.Len(t, leaderPeers, 1)
+		for p, n := range leaderPeers {
+			assert.Equal(t, 3, n, "peer %q expected 3 uses", p)
+		}
+	})
+
+	// Check them all for the bad error
+	const grpcError = `failed to find Consul server for global address`
```

Comment:
This error would periodically show up when executing various API calls that ultimately used gRPC to fulfill them under the covers.